File: dev/test_data/main/golang/expected/aas_core_meta.v3/input/snippets/Verification/data_specification_IEC_61360s_have_definition_at_least_in_english.go
// Check that [aastypes.DataSpecificationIec61360.Definition]
// is defined for all data specifications whose content is given as
// IEC 61360 at least in English.
func DataSpecificationIEC61360sHaveDefinitionAtLeastInEnglish(
	embeddedDataSpecifications []aastypes.IEmbeddedDataSpecification) bool {
	for _, eds := range embeddedDataSpecifications {
		content := eds.DataSpecificationContent()

		ok := aastypes.IsDataSpecificationIEC61360(content)
		if !ok {
			continue
		}
		iec61360 := content.(aastypes.IDataSpecificationIEC61360)

		definition := iec61360.Definition()
		if definition == nil {
			return false
		}

		noEnglish := true
		for _, langString := range definition {
			if IsBCP47ForEnglish(langString.Language()) {
				noEnglish = false
				break
			}
		}

		if noEnglish {
			return false
		}
	}
	return true
}

File: dev/test_data/main/cpp/expected/aas_core_meta.v3/input/snippets/verification/data_specification_IEC_61360s_have_definition_at_least_in_english.cpp
bool DataSpecificationIec61360sHaveDefinitionAtLeastInEnglish(
  const std::vector<
    std::shared_ptr<types::IEmbeddedDataSpecification>
  >& embedded_data_specifications
) {
  for (const auto& embedded_data_specification : embedded_data_specifications) {
    const std::shared_ptr<
      types::IDataSpecificationContent
    >& content(
      embedded_data_specification->data_specification_content()
    );

    // NOTE (mristin):
	// We need to use dynamic cast due to virtual inheritance. Otherwise,
    // we would have used the model_type followed by a static cast.
	const types::IDataSpecificationIec61360* iec61360 = dynamic_cast<
	  const types::IDataSpecificationIec61360*
	>(content.get());

    if (iec61360 != nullptr) {
      const common::optional<
        std::vector<
          std::shared_ptr<types::ILangStringDefinitionTypeIec61360>
        >
      >& maybe_definition = iec61360->definition();

      if (!maybe_definition.has_value()) {
        return false;
      }

      const std::vector<
        std::shared_ptr<types::ILangStringDefinitionTypeIec61360>
      >& definition = *maybe_definition;

      bool no_definition_in_english = true;
      for (const auto& lang_string : definition) {
        if (IsBcp47ForEnglish(lang_string->language())) {
          no_definition_in_english = false;
          break;
        }
      }

      if (no_definition_in_english) {
      	return false;
      }
    }
  }

  return true;
}
